File: src/memento/control/scheduler.py
from __future__ import annotations

import json
import sqlite3
from dataclasses import dataclass
from datetime import datetime, timezone
from uuid import uuid4

from memento.answers import ModelAttempt


@dataclass(frozen=True, slots=True)
class SchedulerRunRecord:
    run_id: str
    job_name: str
    window_key: str
    base_revision: str | None
    end_revision: str | None
    state: str
    signal_count: int
    proposal_count: int
    model_chain: tuple[ModelAttempt, ...]
    started_at: str
    finished_at: str | None
    error_message: str | None


class SchedulerConflictError(RuntimeError):
    """Raised when a scheduler window is already running or completed."""


@dataclass(frozen=True, slots=True)
class SchedulerClaim:
    created: bool
    record: SchedulerRunRecord


def utcnow() -> str:
    return datetime.now(tz=timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def claim_scheduler_run(
    connection: sqlite3.Connection,
    *,
    job_name: str,
    window_key: str,
    base_revision: str | None,
) -> SchedulerClaim:
    running = connection.execute(
        "SELECT * FROM scheduler_runs WHERE job_name = ? AND state = 'running' LIMIT 1",
        (job_name,),
    ).fetchone()
    if running is not None:
        raise SchedulerConflictError(f"job {job_name} is already running")
    existing = connection.execute(
        "SELECT * FROM scheduler_runs WHERE job_name = ? AND window_key = ?",
        (job_name, window_key),
    ).fetchone()
    if existing is not None:
        return SchedulerClaim(created=False, record=_row_to_run(existing))
    run_id = str(uuid4())
    started_at = utcnow()
    with connection:
        connection.execute(
            """
            INSERT INTO scheduler_runs(
                run_id, job_name, window_key, base_revision, state, started_at
            ) VALUES(?, ?, ?, ?, 'running', ?)
            """,
            (run_id, job_name, window_key, base_revision, started_at),
        )
    return SchedulerClaim(created=True, record=get_scheduler_run(connection, run_id))


def get_scheduler_run(connection: sqlite3.Connection, run_id: str) -> SchedulerRunRecord:
    row = connection.execute("SELECT * FROM scheduler_runs WHERE run_id = ?", (run_id,)).fetchone()
    if row is None:
        raise KeyError(run_id)
    return _row_to_run(row)


def finish_scheduler_run(
    connection: sqlite3.Connection,
    run_id: str,
    *,
    state: str,
    end_revision: str | None,
    signal_count: int,
    proposal_count: int,
    model_chain: tuple[ModelAttempt, ...] = (),
    error_message: str | None = None,
) -> SchedulerRunRecord:
    with connection:
        connection.execute(
            """
            UPDATE scheduler_runs
            SET state = ?, end_revision = ?, signal_count = ?, proposal_count = ?,
                model_chain_json = ?, error_message = ?, finished_at = ?
            WHERE run_id = ?
            """,
            (
                state,
                end_revision,
                signal_count,
                proposal_count,
                json.dumps([item.model_dump(mode="json") for item in model_chain]),
                error_message,
                utcnow(),
                run_id,
            ),
        )
    return get_scheduler_run(connection, run_id)


def _row_to_run(row: sqlite3.Row) -> SchedulerRunRecord:
    chain = json.loads(row["model_chain_json"] or "[]")
    attempts = tuple(
        ModelAttempt(model=str(item), outcome="success")
        if isinstance(item, str)
        else ModelAttempt.model_validate(item)
        for item in chain
    )
    return SchedulerRunRecord(
        run_id=row["run_id"],
        job_name=row["job_name"],
        window_key=row["window_key"],
        base_revision=row["base_revision"],
        end_revision=row["end_revision"],
        state=row["state"],
        signal_count=int(row["signal_count"]),
        proposal_count=int(row["proposal_count"]),
        model_chain=attempts,
        started_at=row["started_at"],
        finished_at=row["finished_at"],
        error_message=row["error_message"],
    )

File: go/control/scheduler.go
package control

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"github.com/google/uuid"
	"github.com/rcarmo/memento/go/internal/pyjson"
	"time"
)

type ModelAttempt struct {
	Model   string `json:"model"`
	Outcome string `json:"outcome"`
}
type SchedulerRunRecord struct {
	RunID                      string `json:"run_id"`
	JobName                    string `json:"job_name"`
	WindowKey                  string `json:"window_key"`
	BaseRevision, EndRevision  *string
	State                      string `json:"state"`
	SignalCount, ProposalCount int
	ModelChain                 []ModelAttempt
	StartedAt                  string
	FinishedAt, ErrorMessage   *string
}
type SchedulerClaim struct {
	Created bool
	Record  SchedulerRunRecord
}
type SchedulerConflictError struct{ JobName string }

func (e *SchedulerConflictError) Error() string { return "job " + e.JobName + " is already running" }

type Scheduler struct {
	DB   *sql.DB
	Now  func() time.Time
	UUID func() string
}

func (s Scheduler) now() string {
	now := time.Now()
	if s.Now != nil {
		now = s.Now()
	}
	return now.UTC().Truncate(time.Second).Format(time.RFC3339)
}
func (s Scheduler) id() string {
	if s.UUID != nil {
		return s.UUID()
	}
	return uuid.NewString()
}

const schedulerColumns = "run_id,job_name,window_key,base_revision,end_revision,state,signal_count,proposal_count,model_chain_json,started_at,finished_at,error_message"

func (s Scheduler) Claim(ctx context.Context, job, window string, base *string) (SchedulerClaim, error) {
	if _, err := scanScheduler(s.DB.QueryRowContext(ctx, "SELECT "+schedulerColumns+" FROM scheduler_runs WHERE job_name=? AND state='running' LIMIT 1", job)); err == nil {
		return SchedulerClaim{}, &SchedulerConflictError{job}
	} else if !errors.Is(err, sql.ErrNoRows) {
		return SchedulerClaim{}, err
	}
	if row, err := scanScheduler(s.DB.QueryRowContext(ctx, "SELECT "+schedulerColumns+" FROM scheduler_runs WHERE job_name=? AND window_key=?", job, window)); err == nil {
		return SchedulerClaim{false, row}, nil
	} else if !errors.Is(err, sql.ErrNoRows) {
		return SchedulerClaim{}, err
	}
	id, started := s.id(), s.now()
	if _, err := s.DB.ExecContext(ctx, "INSERT INTO scheduler_runs(run_id,job_name,window_key,base_revision,state,started_at) VALUES(?,?,?,?,'running',?)", id, job, window, base, started); err != nil {
		return SchedulerClaim{}, err
	}
	record, err := s.Get(ctx, id)
	return SchedulerClaim{true, record}, err
}
func (s Scheduler) Get(ctx context.Context, id string) (SchedulerRunRecord, error) {
	return scanScheduler(s.DB.QueryRowContext(ctx, "SELECT "+schedulerColumns+" FROM scheduler_runs WHERE run_id=?", id))
}
func (s Scheduler) Finish(ctx context.Context, id, state string, end *string, signals, proposals int, chain []ModelAttempt, message *string) (SchedulerRunRecord, error) {
	values := make([]any, 0, len(chain))
	for _, attempt := range chain {
		values = append(values, map[string]any{"model": attempt.Model, "outcome": attempt.Outcome})
	}
	raw, _ := pyjson.Dumps(values)
	var err error
	if _, err = s.DB.ExecContext(ctx, "UPDATE scheduler_runs SET state=?,end_revision=?,signal_count=?,proposal_count=?,model_chain_json=?,error_message=?,finished_at=? WHERE run_id=?", state, end, signals, proposals, raw, message, s.now(), id); err != nil {
		return SchedulerRunRecord{}, err
	}
	return s.Get(ctx, id)
}
func scanScheduler(row rowScanner) (SchedulerRunRecord, error) {
	var record SchedulerRunRecord
	var raw *string
	if err := row.Scan(&record.RunID, &record.JobName, &record.WindowKey, &record.BaseRevision, &record.EndRevision, &record.State, &record.SignalCount, &record.ProposalCount, &raw, &record.StartedAt, &record.FinishedAt, &record.ErrorMessage); err != nil {
		return record, err
	}
	record.ModelChain = []ModelAttempt{}
	if raw == nil {
		return record, nil
	}
	var values []any
	if err := json.Unmarshal([]byte(*raw), &values); err != nil {
		return record, err
	}
	for _, value := range values {
		switch item := value.(type) {
		case string:
			record.ModelChain = append(record.ModelChain, ModelAttempt{item, "success"})
		case map[string]any:
			model, modelOK := item["model"].(string)
			outcome, outcomeOK := item["outcome"].(string)
			if !modelOK || !outcomeOK {
				return record, errors.New("invalid scheduler model chain")
			}
			record.ModelChain = append(record.ModelChain, ModelAttempt{model, outcome})
		default:
			return record, errors.New("invalid scheduler model chain")
		}
	}
	return record, nil
}
